plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSegmentPruner.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spooling.filesystem;
+
+import com.google.inject.Inject;
+import io.airlift.log.Logger;
+import io.airlift.units.Duration;
+import io.trino.filesystem.FileEntry;
+import io.trino.filesystem.FileIterator;
+import io.trino.filesystem.Location;
+import io.trino.filesystem.TrinoFileSystem;
+import io.trino.filesystem.TrinoFileSystemFactory;
+import io.trino.spi.security.ConnectorIdentity;
+import jakarta.annotation.PostConstruct;
+import jakarta.annotation.PreDestroy;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ScheduledExecutorService;
+
+import static io.trino.spooling.filesystem.FileSystemSpooledSegmentHandle.fromStorageObjectName;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class FileSystemSegmentPruner
+{
+    private final Logger log = Logger.get(FileSystemSegmentPruner.class);
+
+    private final TrinoFileSystem fileSystem;
+    private final ScheduledExecutorService executor;
+    private final boolean enabled;
+    private final Duration interval;
+    private final Location location;
+    private final long batchSize;
+    private boolean closed;
+
+    private final List<Location> expiredSegments = new ArrayList<>();
```

Comment:
why keep it as a field. It is errorprone as you may call `prune()` while `expiredSegments` is not cleared and add copies to this list. Just build new list on every call.

Reply:
Done.